rfd/0171-database-session-playback.md

```diff
@@ -0,0 +1,345 @@
+can only be viewed as JSON:
+
+
+```json
+[
+  {
+      "ei": 0,
+      "event": "db.session.start",
+      "code": "TDB00I",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 1,
+      "event": "db.session.postgres.statements.execute",
+      "code": "TPG02I",
+      "portal_name": "test",
+      ...
+  },
+  {
+      "ei": 2,
+      "event": "db.session.query",
+      "code": "TDB02I",
+      "db_query": "SELECT 1;",
+      "success": true,
+      ...
+  },
+  {
+      "ei": 3,
+      "event": "db.session.end",
+      "code": "TDB01I",
+      ...
+  }]
+```
+
+### Audit events
+
+To have a complete recording with users input and server response, we'll
+include the command result event:
+
+```proto
+// NOTE: This message already exists, we are just reusing as it is.
+// Copied from: api/proto/teleport/legacy/types/events/events.proto
+message Status {
+  // Success indicates the success or failure of the operation
+  bool Success = 1 [(gogoproto.jsontag) = "success"];
+
+  // Error includes system error message for the failed attempt
+  string Error = 2 [(gogoproto.jsontag) = "error,omitempty"];
+
+  // UserMessage is a user-friendly message for successfull or unsuccessfull auth attempt
+  string UserMessage = 3 [(gogoproto.jsontag) = "message,omitempty"];
+}
+
+// DatabaseSessionCommandResult represents the result of a user command. It is
+// expected that for each user command/query there will be a corresponding
+// result.
+message DatabaseSessionCommandResult {
+  // Metadata is a common event metadata.
+  Metadata Metadata = 1;
+  // User is a common user event metadata.
+  UserMetadata User = 2;
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3;
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4;
+  // Status of the execution.
+  Status status = 5;
+  // AfftectedRecords represents the number of records that were affected by the
+  // user query.
+  uint64 AffectedRecords = 6;
+}
+```
+
+#### Examples
+
+* Audit event generated for a query that returns data (`SELET id, name FROM events;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true
+  },
+  "affected_records": 5
+}
+```
+
+* Query that only returns a success message (`UPDATE events SET ...`):
+
+```json
+{
+  ...
+  "status": {
+    "success": true
+  },
+  "affected_records": 3
+}
+```
+
+* Query that returns an error (`SELECT err;`):
+
+```json
+{
+  ...
+  "status": {
+    "success": false,
+    "error": "ERROR: column \"err\" does not exist"
+  },
+  "affected_records": 0
+}
+```
+
+### Recording options
+
+Database session recording options will be defined per role. This configuration
+will be under the `options.record_session.db` role option. It will support the
+following values:
+- `on`: Enables session recording.
+- `off`: Disables session recording. Audit events are kept unchanged (start,
+  end, and query events emitted).
```

Comment:
As far as I recall, all database access events currently go to both audit log and session recording. So we're just adding an ability to replay them (and an extra event for query result). Is that correct? Will `off` option basically mean that we stop emitting events to session recordings?

Reply:
> As far as I recall, all database access events currently go to both audit log and session recording. So we're just adding an ability to replay them (and an extra event for query result). Is that correct?

Yes.

> Will off option basically mean that we stop emitting events to session recordings?

Yes. This change will not affect audit events (similar to SSH enhanced session recordings). Currently, there is no option to do this other than completely disabling the session recordings on the cluster level (`auth_server.session_recording: "off").